attachdetach controller: attach volumes immediately when Pod's PVCs are bound

cofyc opened the pull request:
**What this PR does / why we need it**:

Let attachdetach controller to attach volumes immediately when Pod's PVCs are bound.

Current attachdetach controller calls `util.ProcessPodVolume` to add pod volumes into `desiredStateOfWorld` on these events:

- podAdd event
- podUpdate event
- podDelete event
- periodical `desiredStateOfWorldPopulator.findAndAddActivePod`

But if a pod is created with PVCs not bound, no volumes will be added into `desiredStateOfWorld` [because PVCs not bound](https://github.com/kubernetes/kubernetes/blob/v1.12.0-alpha.0/pkg/controller/volume/attachdetach/util/util.go#L99). When pv controller binds PVCs successfully, attachdetach controller will not add pod volumes immediately because it does not watch on PVC events.

It will wait until a pod update event is triggered (normally will not happen because no new status will be reported by kubelet) or `desiredStateOfWorldPopulator.findAndAddActivePod` is called (maybe 0~3 minutes later, see [timer configs](https://github.com/kubernetes/kubernetes/blob/v1.12.0-alpha.0/pkg/controller/volume/attachdetach/attach_detach_controller.go)).

In bad case, pod start time will be very long (~3 minutes + ~2 minutes (kubelet max exponential backoff)), for example: https://github.com/kubernetes/kubernetes/issues/64549#issuecomment-409440546.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #64549

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
attachdetach controller attaches volumes immediately when Pod's PVCs are bound
```

gnufied commented on pkg/controller/volume/attachdetach/populator/desired_state_of_world_populator.go:
```diff
@@ -168,3 +170,40 @@ func (dswp *desiredStateOfWorldPopulator) findAndAddActivePods() {
+		// Skip unbound PVCs.
+		return
+	}
+	pods, err := dswp.podLister.List(labels.Everything())
```
Another way - perhaps we can solve this is, by implementing work queues in attach/detach controller. The problem here is - any random PVC update will cause iteration of all pods in the cluster. One can perhaps DOS the cluster by constantly updating the pvc?

cofyc replied:
Yes, you're right. I'll update it.

cofyc replied:
Because PVC does not contains Pod references, but we can index pods by PVC locally, then we don't need to iterate all pods every time. For example, daemon controller [indexes pods based on their NodeName](https://github.com/kubernetes/kubernetes/blob/v1.12.0-alpha.1/pkg/controller/daemon/daemon_controller.go#L197).

I think this will improve performance in large cluster.

cofyc commented:
/assign @saad-ali 
Ready to review, summary of changes:

- Use queue to process PVCs on add/update events
- Index pods by PVC key then we don't need to iterate to find pods
- Add integration test for this feature

cc @jingxu97 @msau42

cofyc commented:
/test pull-kubernetes-integration
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-e2e-kops-aws

redbaron commented:
@cofyc , do you think it is cherry-pickable to 1.10, 1.11?

msau42 commented:
@kubernetes/sig-storage-pr-reviews

gnufied commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -179,15 +184,53 @@ func NewAttachDetachController(
+		UpdateFunc: func(old, new interface{}) {
+			adc.enqueuePVC(new)
+		},
+		DeleteFunc: nil,
```
I dont think you need to specify `DeleteFunc` if you don't want to.

cofyc replied:
Updated.

gnufied reviewed (commented):
mostly looks good to me.

cofyc commented:
@redbaron 
Patch of `attachdetach` is easy to be cherry-picked, test part have some conflicts. 
I'll try to cherry-pick it into old releases when it's accepted.

gnufied commented:
@cofyc can you please squash your commits?

Commit:
attachdetach controller: attach volumes immediately when Pod's PVCs are bound
- Use queue to process PVCs on add/update events
- Index pods by PVC key then we don't need to iterate to find pods

Commit:
attachdetach controller: attach volumes immediately when Pod's PVCs are bound
- Add integration test for this feature

cofyc commented:
@gnufied 
Done. I split it into two parts:

- attachdetach controller: https://github.com/kubernetes/kubernetes/pull/66863/commits/8db47db84a09e3fa00c240383519f6a19ffa200a 
- integration tets: https://github.com/kubernetes/kubernetes/pull/66863/commits/83366218a0eb990b73c1f180bf923f270e5c4be2

jingxu97 commented on test/integration/volume/attach_detach_test.go:
```diff
@@ -446,3 +523,85 @@ func TestPodAddedByDswp(t *testing.T) {
+	waitForPodFuncInDSWP(t, ctrl.GetDesiredStateOfWorld(), 10*time.Second, "expected 1 pod in dsw", 1)
+	createPVForPVC(t, testClient, pvc)
+	waitForPodFuncInDSWP(t, ctrl.GetDesiredStateOfWorld(), 20*time.Second, "expected 2 pods in dsw after PVCs are bound", 2)
+	close(stopCh)
```
I wonder how well 10s and 20s setting will work? What is the typical delay under normal load, and heavy load?

cofyc replied:
attachdetach controller processes Pod's PVs to attach very fast, the delay should be almost same as apiserver watch event delay, because it does nothing except adding Pod's Volumes into DSWP (no api calls) on event. 
In my local test, delay to attach PV is less than 100ms. The most delay is waiting PV controller to bound PVC, normally 4~5 seconds.
Perhaps we should increase `timeout` of `waitForPodFuncInDSWP`, and it will not increase execution time of test.
`time.Sleep(10 * time.Second)` is a hack to make sure the second pod cannot be added into DSWP at least in given seconds, because no object state or event can be checked to verify. 10 seconds should be enough here.

cofyc replied:
`timeout` of `waitForPodFuncInDSWP` is increased to 60 seconds now.

msau42 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -179,15 +184,52 @@ func NewAttachDetachController(
+	if !ok {
+		return []string{}, nil
+	}
+	// We are only interested in active pods with nodeName set
```
Can you also add a comment that this cache is only used for attaching

cofyc replied:
Updated.

msau42 reviewed (commented):
This is awesome! I just have two nits

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -486,6 +534,83 @@ func (adc *attachDetachController) nodeDelete(obj interface{}) {
+func (adc *attachDetachController) enqueuePVC(obj interface{}) {
+	key, err := kcache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", obj, err))
```
Just curious: Why `runtime.HandleError`? Why not just `glog.Errorf`?

cofyc replied:
I followed https://github.com/kubernetes/kubernetes/blob/v1.12.0-alpha.1/pkg/controller/daemon/daemon_controller.go#L316, and found [sample-controller](https://github.com/kubernetes/sample-controller/blob/master/controller.go#L341) uses `runtime.HandleError` to log errors too. Perhaps because `runtime.HandleError` is dedicated to log non-user facing ignored errors and can be extended by adding error handlers.

saad-ali commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -486,6 +534,83 @@ func (adc *attachDetachController) nodeDelete(obj interface{}) {
+		return nil
+	}
+
+	objs, err := adc.podIndexer.ByIndex(pvcKeyIndex, key)
```
Just to verify, if there are multiple pods using the same PVC, this will return all pods? If so make sure there is a test case for that.

cofyc replied:
Yes, test has been updated.

saad-ali reviewed (commented):
Largely LGTM, a couple of questions.

cofyc commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -486,6 +534,83 @@ func (adc *attachDetachController) nodeDelete(obj interface{}) {
+	}
+	pvc, err := adc.pvcLister.PersistentVolumeClaims(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		glog.V(2).Infof("error getting pvc %q from informer: %v", key, err)
```
Fixed, increased to log level 4.

cofyc commented:
Squashed into two commits:

- attachdetach controller: 5fafae11d8dc33ae3120f2b686d762d4101d4ecc
- integration test: 17ab4c36ce0c012e0447f338315307d6db5b3ac8

msau42 commented:
/test pull-kubernetes-integration

redbaron commented:
/test pull-kubernetes-integration

redbaron commented:
/retest

cofyc commented:
@redbaron 

I've created cherry pick PRs for 1.9/1.10/1.11:

- release-1.9: https://github.com/kubernetes/kubernetes/pull/67417
- release-1.10: https://github.com/kubernetes/kubernetes/pull/67415
- release-1.11: https://github.com/kubernetes/kubernetes/pull/67413
